Comparison of Abel Transform Methods

MikhailRyazanov opened the pull request:
Let's start filling the “Comparison of Abel Transform Methods” section with useful information. I have made the plots for “Speed benchmarks” (see #280) and wrote some “general advice” about libraries and using “daz” (see #246).

This is a PR from the same repository. Let's see how this works....
If I understood [this collaborative workflow](https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_private_team) correctly, everybody will need to do `git fetch upstream` and then checkout the `comparison` (or `upstream/comparison`?) branch. Then everybody can make commits to it, but will need to do `git pull` (or `git fetch` followed by `git merge`) to synchronize the local copy with other people's changes before pushing new changes.

The speed benchmarks data and plotting scripts are currently placed in `doc/benchmarks`. If we also want to compare transform results, maybe it would be better to move this stuff to `doc/benchmarks/speed` or `doc/benchmarks/timings` and create another subdirectory in `doc/benchmarks` for the transform results. Or maybe place them in a separate directory `doc/comparison` or something like that.

The help that I need for this part from @stggh is to add his script that generates the timings and to update the data files with his new timings (for i7-9700).

And @DanHickstein also wanted:

> I would discuss the quality of the transform results first and then move on to the speed benchmarks. Probably we should simply adapt our manuscript to fit on this page and keep a "living version" of the comparison paper here.

so please go ahead. :–)

(I think, if we discuss proposed changes here and then commit them one by one instead of concurrently, this should minimize possible edit conflicts.)

DanHickstein commented:
Thanks for getting started on this Mikhail!

Is there any advantage of using commits to directly work on the same branch rather than making PRs to this branch?

I will get started working on including some more of the information from the paper into the comparison document.

I'm going to set up doc/comparison as a place where all of the scripts and data for making the figures for the comparison article can live.

MikhailRyazanov commented:
> Is there any advantage of using commits to directly work on the same branch rather than making PRs to this branch?

Only less work and cleaner history (no useless intermediate merge commits). PRs are intended for merging outside contributions (from people who cannot commit directly) or for merging “features” into master, but routine collaboration within a repository is more naturally done directly in “feature branches”.

> I'm going to set up doc/comparison as a place where all of the scripts and data for making the figures for the comparison article can live.

OK, then no directory reorganization is needed. @stggh's script will go into doc/benchmarks, with the updated data.

pep8speaks commented:
Hello @MikhailRyazanov! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`doc/benchmarks/pyabel-plot.py`](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py):

> [Line 13:18](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L13): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 14:18](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L14): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 15:18](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L15): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 16:18](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L16): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 17:18](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L17): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 19:18](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L19): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 20:18](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L20): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 22:18](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L22): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 23:18](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L23): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 24:18](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L24): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 27:26](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L27): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 27:28](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L27): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 27:42](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L27): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 34:5](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L34): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'
> [Line 40:5](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L40): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 47:1](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L47): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 48:1](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L48): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 56:80](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L56): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 56:92](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L56): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\;'
> [Line 57:80](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L57): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 57:92](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L57): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\;'
> [Line 58:80](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L58): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 58:92](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L58): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\;'
> [Line 59:80](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L59): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 59:92](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L59): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\;'
> [Line 62:47](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L62): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 73:80](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L73): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 79 characters)
> [Line 73:98](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L73): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 78:80](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L78): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 79 characters)
> [Line 79:9](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L79): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 90:1](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L90): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 92:50](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L92): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 93:80](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L93): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (122 > 79 characters)
> [Line 95:1](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L95): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 96:1](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L96): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 98:1](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L98): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 100:1](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-plot.py#L100): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`doc/benchmarks/pyabel-timing.py`](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-timing.py):

> [Line 68:28](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-timing.py#L68): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 70:14](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-timing.py#L70): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 102:22](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-timing.py#L102): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 137:8](https://github.com/PyAbel/PyAbel/blob/549e10381f24710f9f967715b7d436ff4f27ffe3/doc/benchmarks/pyabel-timing.py#L137): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

stggh commented:
Okay, the `comparison` branch makes sense, apart from renegade @DanHickstein making his own PR #291 ;-) .

I have pushed some i7_9700 timings to the `comparison` branch. Note, my code hack to include `rbasex` and `rbasex(None)` into `pyabel-timing.py` may not be the best.  I will fix the pep8 issues.

Sorry, for the delay, I wanted to check that the old i7_6700 timings were similar to what we had in the past. Also, my wife has me painting rooms in our house ...

DanHickstein commented:
> apart from renegade @DanHickstein making his own PR

I won't conform to your laws! :)

> I have pushed some i7_9700 timings to the comparison branch.

Awesome!

> Also, my wife has me painting rooms in our house 

Ha! I just repainted one of our bedrooms a few weeks ago. Painting always takes a lot more time that I think.

DanHickstein commented:
The last point of rBasex (and rBasex (None)) seems like an outlier. Are we running into RAM limitations or something? I'm tempted to not plot it...
![image](https://user-images.githubusercontent.com/1107796/80557341-d0881b00-8993-11ea-9747-7b31266f2418.png)

stggh commented:
May well be, I am re-running.

stggh commented:
Raspberry pi 4B benchmark:
<img width="337" alt="Screen Shot 2020-05-02 at 10 30 38" src="https://user-images.githubusercontent.com/10932229/80850605-2690e300-8c60-11ea-8b16-a87a0f1f390e.png">

MikhailRyazanov commented:
Well, I have merged @DanHickstein's #291 here...

A question to everybody:
Do we really need the timings files for several different machines if we do not use them at all? Maybe they should be removed, leaving only the actual ones, as I had initially? Or, alternatively, they all can be plotted as well. In tat case, my suggestion would be to move the phrase “These benchmarks were completed using a personal computer equipped with a 3.0 GHz Intel i7-9700 processor and 32 GB RAM running GNU/Linux.” from Fig. 6 caption to the first paragraph of “Speed benchmarks”, since this remark is really about all the plots there, and after that sentence give links to subpages with the same plots for other machines. (Then I can also provide benchmarks for the cheap AMD Ryzen under Windows (and Linux), to cover a broader range of environments.)

DanHickstein commented:
> Do we really need the timings files for several different machines if we do not use them at all? Maybe they should be removed, leaving only the actual ones, as I had initially?

I don't feel strongly about this. We can remove the old ones and just include the most recent ones if you like. Unless there is something really interesting about showing benchmarks from multiple different processors, I would suggest that we just show benchmarks from one processor. I think that the point is to just show the general scaling and speed differences between the methods, not to get too far into the nitty gritty details about different processors.

MikhailRyazanov commented:
Thanks, @DanHickstein. I also thought initially that one set of benchmarks would be enough and asked @stggh to just replace the old files with his new results, but he decided to keep the old ones too, and even added more benchmarks.

These Raspberry Pi results are actually interesting in that rBasex becomes the most efficient method for all practical image sizes. :–) I guess, this is only due to the worse matrix-algebra performance, and perhaps nobody will run PyAbel on Raspberry Pi, but nonetheless...

My previous tests on AMD Ryzen also showed that speed differences between the methods are not always the same as in Steve's results (this depends on the number of cores and the balance between floating-point, integer, flow-control and memory-access operations); there were also differences between Linux and Windows on the same machine. Another consideration of more practical importance is that showing only the results from this expensive high-performance hardware can make an impression that people really need it to use PyAbel, although in reality it has good performance even on much cheaper hardware.

So, removing the unused data is the easiest option, but on the other hand, showing that data also has some advantages.

MikhailRyazanov commented:
By the way, @stggh,

>> The last point of rBasex (and rBasex (None)) seems like an outlier. Are we running into RAM limitations or something? I'm tempted to not plot it...
>
> May well be, I am re-running.

Did you get anything from that? I wouldn't be surprised if it starts swapping (which you should be able to see from RAM/swap usage) at such image sizes, unreasonably large for VMI data, since I did not try to minimize memory usage at the expense of speed for typical image sizes. If this is the case, maybe we indeed should not plot these last points for rBasex and Hansen–Law.

stggh commented:
No, the results are the same (sorry, I thought that I posted this, before the Rpi results):
```
# rbasex          iabel(ms)           (None)(ms)          basis(ms)
    5             0.11261             0.06864             0.10807
    9             0.11310             0.06925             0.14842
   17             0.11511             0.06955             0.22544
   33             0.12354             0.07422             0.38970
   65             0.14372             0.08550             0.93613
  129             0.21301             0.12419             1.74566
  257             0.59298             0.33920             3.23757
  513             1.83382             0.97952             7.80618
 1025             8.30427             4.52088            20.75333
 2049            36.38921            18.02486            51.86263
 4097           175.62474            90.95619           221.65900
 8193           696.05620           344.29510          1017.67079
16385          2694.15025          1254.63244          5917.78351
32769        146544.18560         43537.52580         37277.30449
```
```
# hansenlaw       iabel(ms)           
    5             0.11626
    9             0.13581
   17             0.17827
   33             0.26336
   65             0.47840
  129             1.02095
  257             2.60069
  513             8.23860
 1025            33.69837
 2049           115.44571
 4097           489.75560
 8193          1557.08233
16385          6265.46319
32769         31533.56482
```
![image](https://user-images.githubusercontent.com/10932229/81629152-d7a83200-9445-11ea-98dd-c4b53e817b3b.png)

I think the issue is not memory, but thermal control. `dmesg` gives:
![image](https://user-images.githubusercontent.com/10932229/81639332-f9fa7980-945e-11ea-9e9d-6629e680c834.png)
`sensors` does not detect any fan, so I am not sure how to control the cooling outside of the BIOS. I will take a look later.

>  I also thought initially that one set of benchmarks would be enough and asked @stggh to just replace the old files with his new results, but he decided to keep the old ones too, and even added more benchmarks.

Yes, but I did not like the idea of stomping over the old results. It is also "risky" to rely on one PC for the benchmarks, without verification from another similar system.

>  speed differences between the methods are not always the same

I favour benchmarks from various systems, but am also ambivalent.

MikhailRyazanov commented:
Thanks, @stggh! I guess, the CPU fan has automatic control, so it does speed up when the temperature rises, thus it's unlikely that the observed throttling can be overcome...

If Dan has no objections, I'll try to create an auxiliary page with all the benchmarks from various systems, and if we don't like it, it can be removed.
But I will need the `rbasex` and `rbasex(None)` benchmarks from the old i7-6700 system. If you have them, please post them here (or replace my i5 data, which I'm not going to use anyway, in `doc/transform_methods/comparison/fig_benchmarks/benchmarks_2019/`).

stggh commented:
Oddly, the fan speed/noise does not appear to increase, so something is not right with the fan control. I haven't yet rebooted to check the BIOS settings.

>  benchmarks from the old i7-6700 system
```
benchmarks_i7-6700_Linux_5.3.0-40-generic
```
no high temperature warning in dmesg.
```
 rbasex          iabel(ms)           basis(ms)
    5             0.10452             0.10355
    9             0.10721             0.14982
   17             0.11027             0.24400
   33             0.12008             0.42883
   65             0.14302             0.92872
  129             0.21891             1.72293
  257             0.59280             3.44470
  513             2.07769             8.50229
 1025             8.82631            24.83911
 2049            41.96303            66.59847
 4097           222.23797           292.11110
 8193           884.57199          1472.04311
16385          3275.78789          8700.22730
32769        217201.64723         58130.10464
```
```
# rbasex(None)    iabel(ms)           
    5             0.05711
    9             0.05883
   17             0.06006
   33             0.06565
   65             0.07777
  129             0.12134
  257             0.32868
  513             1.09392
 1025             4.80125
 2049            20.06163
 4097           118.22915
 8193           487.11763
16385          1653.54199
32769         45508.45924
```

stggh commented:
Dell Optiplex 7070 BIOS only has the option to turn off fan control, and run the fan flat out. Sadly, the timing is similar to before:
```
# rbasex          iabel(ms)           basis(ms)
 2049            36.15992            50.07761
 4097           177.16946           214.18520
 8193           702.31928           999.62982
16385          2674.88479          5587.29921
32769         84370.10413         35879.46093
```
In this case there was no `dmesg` CPU throttling. Hence, it looks as though the issue is, after all, insufficient RAM.

DanHickstein commented:
> If Dan has no objections, I'll try to create an auxiliary page with all the benchmarks from various systems, and if we don't like it, it can be removed.

Yeah, that sounds like fun. I doubt that it will be of too much interest to the general userbase, but it sounds like fun for us geeks. So why not? But, we shouldn't use up too much of our energy on benchmarks. I think that we tend to get a little carried away :). (That said, some of our big speed improvements have resulted from our tendency to run lots of comparative benchmarks.)

MikhailRyazanov commented:
Thanks, @stggh! The i7-6700 `rbasex` data now completes that dataset.
> the timing is similar to before:

Well, this:
```
# rbasex          iabel(ms)           basis(ms)
...
32769         84370.10413         ...
```
is noticeably better than before:
```
32769        133686.80234         ...
```
What is the last line for `rbasex(None)`?

> Odd that ... systems behave differently.

My AMD under Win10 behaves even more <s>oddly</s> differently!
![tt](https://user-images.githubusercontent.com/12854133/81818303-1ba14280-94eb-11ea-9598-f170b9d67ee2.png)
![tp](https://user-images.githubusercontent.com/12854133/81818318-1e039c80-94eb-11ea-82e3-2e0d8c14dfca.png)
![bt](https://user-images.githubusercontent.com/12854133/81818326-2065f680-94eb-11ea-8b89-1e651026268a.png)
See how strangely `hansenlaw` does not become quadratic (while `rbasex` remains quadratic)! And how `onion_peeling` basis-set generation behaves for small image sizes...

The `onion_peeling` results are reproducible (I have repeated the test 3 times, separately and with methods in different order). I'll redo the `hansenlaw` to make sure, but it look very systematic. I also want to do this under Linux to see whether it behaves differently (NumPy under Windows definitely has some memory-access issues).

@DanHickstein,
> I doubt that it will be of too much interest to the general userbase,

The unexpected AMD/Win results were interesting to myself at least! :–)

> But, we shouldn't use up too much of our energy on benchmarks.

I already have the benchmarks page, so I'll just need to wait for my benchmarks to finish (while I'm busy with other things).

stggh commented:
```
(py38) [Gibbo:.+.4.0-29-generic]$ cat rbasex.dat
# rbasex          iabel(ms)           basis(ms)
    5             0.09292             0.09155
    9             0.09403             0.13238
   17             0.09617             0.21074
   33             0.10435             0.36328
   65             0.12438             0.87740
  129             0.19134             1.62683
  257             0.56183             3.24362
  513             1.74804             7.61616
 1025             8.01388            19.99236
 2049            36.15992            50.07761
 4097           177.16946           214.18520
 8193           702.31928           999.62982
16385          2652.63505          5717.84295
32769         81248.37484         36181.63510
```
```
(py38) [Gibbo:.+.4.0-29-generic]$ cat rbasex\(None\).dat 
# rbasex(None)    iabel(ms)           
    5             0.05015
    9             0.05055
   17             0.05127
   33             0.05584
   65             0.06696
  129             0.10483
  257             0.31532
  513             0.94748
 1025             4.52871
 2049            18.24100
 4097            95.64268
 8193           358.04784
16385          1247.35140
32769         24476.19363
```

I also find this study interesting.

MikhailRyazanov commented:
Thanks, @stggh! I've updated the i7-9700 timings with your new data.

I also did some corrections to your timing script (it tried to add `(None)` to `linbasex` and thus was crashing) and made it always use a predictable `working` directory instead of `cpuinfo` output (which depends on the kernel revision version...). The timing-data directories are thus also renamed to shorter and more reasonable.

Then I've added my benchmarks on AMD Ryzen ran under Windows and Linux (not complete yet) and made an auxiliary page with all the speed benchmarks plotted. I think, it looks interesting. To complete it, I need to wait for my Linux benchmarks to finish, but I also need some information about @stggh's systems:
* How much RAM this Raspberry Pi has?
* What are the RAM types/speeds for the i7 systems (both)? This can be queried by `sudo dmidecode --type 17`.
* Which Linux distribution and libraries were used on the i7 systems and the RPi? NumPy and SciPy versions can be found in their `__version__` strings, and the BLAS type is probably known (I suppose, MKL on i7, but what on RPi?), however, finding its version depends on how it is installed.

stggh commented:
> made an auxiliary page with all the speed benchmarks plotted

How may I see this page?  Could you post the figure?

**Rpi 4B**
```python
[Rpi4Bp:/home/stg121]$ sudo inxi --system --cpu --memory
X11 connection rejected because of wrong authentication.
System:
  Host: Rpi4Bp Kernel: 4.19.115-v7l+ armv7l bits: 32 Desktop: LXDE 
  Distro: Raspbian GNU/Linux 10 (buster) 
Memory:
  RAM: total: 3.89 GiB used: 710.3 MiB (17.8%) gpu: 256.0 MiB 
  RAM Report: unknown-error: dmidecode was unable to generate data 
Argument "Raspberry Pi 4 Model B Rev 1.1" isn't numeric in sprintf at /usr/bin/inxi line 6969.
CPU:
  Topology: Quad Core model: ARMv7 v7l variant: cortex-a72 bits: 32 
  type: MCP 
  Speed: 1500 MHz min/max: 600/1500 MHz Core speeds (MHz): 1: 1500 2: 1500 
  3: 1500 4: 1500 
```
```
[Rpi4Bp:/home/stg121]$ ipython3
Python 3.7.3 (default, Dec 20 2019, 18:57:59) 

In [1]: import numpy, scipy                                                     

In [2]: numpy.__version__                                                       
Out[2]: '1.16.2'

In [3]: scipy.__version__                                                       
Out[3]: '1.1.0'

In [4]: numpy.show_config()                                                     
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
  NOT AVAILABLE
atlas_3_10_blas_threads_info:
  NOT AVAILABLE
atlas_3_10_blas_info:
  NOT AVAILABLE
atlas_blas_threads_info:
  NOT AVAILABLE
atlas_blas_info:
  NOT AVAILABLE
accelerate_info:
  NOT AVAILABLE
blas_info:
    libraries = ['blas', 'blas']
    library_dirs = ['/usr/lib/arm-linux-gnueabihf']
    include_dirs = ['/usr/local/include', '/usr/include']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    define_macros = [('NO_ATLAS_INFO', 1), ('HAVE_CBLAS', None)]
    libraries = ['blas', 'blas']
    library_dirs = ['/usr/lib/arm-linux-gnueabihf']
    include_dirs = ['/usr/local/include', '/usr/include']
    language = c
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
  NOT AVAILABLE
openblas_clapack_info:
  NOT AVAILABLE
atlas_3_10_threads_info:
  NOT AVAILABLE
atlas_3_10_info:
  NOT AVAILABLE
atlas_threads_info:
  NOT AVAILABLE
atlas_info:
  NOT AVAILABLE
lapack_info:
    libraries = ['lapack', 'lapack']
    library_dirs = ['/usr/lib/arm-linux-gnueabihf']
    language = f77
lapack_opt_info:
    define_macros = [('NO_ATLAS_INFO', 1), ('HAVE_CBLAS', None)]
    libraries = ['lapack', 'lapack', 'blas', 'blas']
    library_dirs = ['/usr/lib/arm-linux-gnueabihf']
    language = c
    include_dirs = ['/usr/local/include', '/usr/include']
```
**i7-9700**
```python
(py38) [Gibbo:/home/stg121]$ sudo inxi --system --cpu --memory
System:
  Host: Gibbo Kernel: 5.4.0-29-generic x86_64 bits: 64 Desktop: Gnome 3.36.1 
  Distro: Ubuntu 20.04 LTS (Focal Fossa) 
Memory:
  RAM: total: 31.15 GiB used: 3.56 GiB (11.4%) 
  Array-1: capacity: 64 GiB slots: 4 EC: None 
  Device-1: DIMM1 size: 16 GiB speed: 2666 MT/s 
  Device-2: DIMM2 size: 16 GiB speed: 2666 MT/s 
  Device-3: DIMM3 size: No Module Installed 
  Device-4: DIMM4 size: No Module Installed 
CPU:
  Topology: 8-Core model: Intel Core i7-9700 bits: 64 type: MCP 
  L2 cache: 12.0 MiB 
  Speed: 2400 MHz min/max: 800/4700 MHz Core speeds (MHz): 1: 2400 2: 2400 
  3: 2400 4: 2400 5: 2400 6: 2400 7: 2400 8: 2400 

```
```
py38) [Gibbo:/home/stg121]$ ipython
Python 3.8.2 (default, Mar 26 2020, 15:53:00)

In [1]: import numpy, scipy                                                    

In [2]: numpy.__version__                                                      
Out[2]: '1.18.1'

In [3]: scipy.__version__                                                      
Out[3]: '1.4.1'

In [4]: numpy.show_config()                                                    
blas_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/envs/py38/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/envs/py38/include']
blas_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/envs/py38/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/envs/py38/include']
lapack_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/envs/py38/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/envs/py38/include']
lapack_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/envs/py38/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/envs/py38/include']
```

**i7-6700**
```python
[etalon:/home/stg121]$ sudo inxi --system --cpu --memory
X11 connection rejected because of wrong authentication.
System:
  Host: etalon Kernel: 5.3.0-40-generic x86_64 bits: 64 Console: tty 6 
  Distro: Ubuntu 19.10 (Eoan Ermine) 
Memory:
  RAM: total: 31.23 GiB used: 5.46 GiB (17.5%) 
  Array-1: capacity: 64 GiB slots: 4 EC: None 
  Device-1: DIMM1 size: 8 GiB speed: 2133 MT/s 
  Device-2: DIMM2 size: 8 GiB speed: 2133 MT/s 
  Device-3: DIMM3 size: 8 GiB speed: 2133 MT/s 
  Device-4: DIMM4 size: 8 GiB speed: 2133 MT/s 
CPU:
  Topology: Quad Core model: Intel Core i7-6700 bits: 64 type: MT MCP 
  L2 cache: 8192 KiB 
  Speed: 800 MHz min/max: 800/4000 MHz Core speeds (MHz): 1: 800 2: 801 
  3: 800 4: 800 5: 800 6: 800 7: 800 8: 800 
```
```
[etalon:/home/stg121]$ ipython3
Python 3.8.2 | packaged by conda-forge | (default, Apr 16 2020, 18:04:51) 

In [1]: import numpy, scipy                                                    

In [2]: numpy.__version__                                                      
Out[2]: '1.18.1'

In [3]: scipy.__version__                                                      
Out[3]: '1.4.1'

In [4]: numpy.show_config()                                                    
blas_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/include']
blas_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/include']
lapack_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/include']
lapack_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/include']
```

MikhailRyazanov commented:
> How may I see this page? Could you post the figure?

There are many figures...
The page itself is at https://pyabel--283.org.readthedocs.build/en/283/transform_methods/comparison/fig_benchmarks/benchmarks.html and is linked form https://pyabel--283.org.readthedocs.build/en/283/transform_methods/comparison.html#speed-benchmarks.

> Rpi 4B

OK, I see: 4 GB RAM, Raspbian GNU/Linux 10, NumPy 1.16.2, SciPy 1.1.0.
Which BLAS implementation does it use? If you list
`ls /usr/lib/arm-linux-gnueabihf/*blas*`
and/or
`apt list --installed "*blas*"`
this will probably clarify.

> i7-9700
> i7-6700

OK, both apparently use DDR4 RAM; OS and NumPy/SciPy are clear, but which MKL versions? (Here I don't know hoe to check.)

stggh commented:
Yes, difficult to find version information.

> Rpi  BLAS
```
libblas3/stable,now 3.8.0-2 armhf [installed,automatic]
```
> which MKL versions

**9700**
```
export MKL_VERBOSE=1
```
```
In [1]: import numpy as np                                                     
mkl-service + Intel(R) MKL: THREADING LAYER: (null)
mkl-service + Intel(R) MKL: setting Intel(R) MKL to use INTEL OpenMP runtime
mkl-service + Intel(R) MKL: preloading libiomp5.so runtime
MKL_VERBOSE Intel(R) MKL 2020.0 Product build 20191122 for Intel(R) 64 architecture Intel(R) Advanced Vector Extensions 2 (Intel(R) AVX2) enabled processors, Lnx 3.00GHz lp64 intel_thread
MKL_VERBOSE SDOT(2,0x55b58db96030,1,0x55b58db96030,1) 22.66us CNR:OFF Dyn:1 FastMM:1 TID:0  NThr:8
```
**6700**
```
In [1]: import numpy as np                                                    
mkl-service + Intel(R) MKL: THREADING LAYER: (null)
mkl-service + Intel(R) MKL: setting Intel(R) MKL to use INTEL OpenMP runtime
mkl-service + Intel(R) MKL: preloading libiomp5.so runtime
MKL_VERBOSE Intel(R) MKL 2019.0 Update 5 Product build 20190808 for Intel(R) 64 architecture Intel(R) Advanced Vector Extensions 2 (Intel(R) AVX2) enabled processors, Lnx 3.40GHz lp64 intel_thread
MKL_VERBOSE SDOT(2,0x55cef4ccd020,1,0x55cef4ccd020,1) 17.63us CNR:OFF Dyn:1 FastMM:1 TID:0  NThr:4`
```

MikhailRyazanov commented:
OK, so “MKL 2019 Update 5” and “MKL 2020 Initial Release” is apparently how they are [officially called](https://software.intel.com/content/www/us/en/develop/articles/intel-math-kernel-library-release-notes-and-new-features.html)...

Regarding RPi, libblas3 is the reference implementation, installed by default. It is not parallelized and does not use advanced instructions, so at least on amd64 it was very slow (see https://github.com/PyAbel/PyAbel/issues/230#issuecomment-433796009 and further discussion there with @rth's suggestions). If you really want, try libopenblas and/or libatlas3 instead — they are available for armhf, although I don't know about their performance. According to [Debian wiki](https://wiki.debian.org/DebianScience/LinearAlgebraLibraries), they all can be installed together and switched using `update-alternatives`, but I was just uninstalling the default libblas3 and liblapack3 libraries to make sure. And yes, testing just a single matrix-heavy method, like `two_point`, should be enough to see the differences.

rth commented:
If at all possible, it's really easy to switch BLAS backend when installing packages from the conda-forge conda channel,
https://conda-forge.org/docs/maintainer/knowledge_base.html#switching-blas-implementation

They have numpy, scipy builds for aarch64 and ppc64le, but not armhf I think.

MikhailRyazanov commented:
I'm done with my benchmarks on AMD Ryzen 3, both Linux and Windows: https://pyabel--283.org.readthedocs.build/en/283/transform_methods/comparison/fig_benchmarks/benchmarks.html#amd-ryzen-3-2200g-linux

Interesting findings from these AMD results:
1. PyAbel under Linux is generally faster than under Windows. I think, mostly due to Python itself, but also due to poorly compiled NumPy for Windows (see https://github.com/numpy/numpy/issues/15513, for example).
2. `hansenlaw`, `three_point` and `two_point` failed at _n_ = 32769 with `MemoryError` under Linux, but were able to complete (not very badly) under Windows. However, `onion_peeling` worked fine for _n_ = 32769 in both cases.
3. `hansenlaw` under Windows behaves very strangely and poorly (I did rerun it, and these are the best results), but under Linux runs on the cheap Ryzen 3 2200G nearly as fast as on i7-9600 and definitely faster than on i7-6700.

@DanHickstein and @stggh, should we merge this now? If Steve wants to play with Raspbery Pi, but not right now, its benchmarks can be updated later.

stggh commented:
I agree, good to merge. Interesting that an old AMD can keep up with modern systems.

On the Rpi I installed openblas, but it made little difference to the `PyAbel` benchmarks:
![image](https://user-images.githubusercontent.com/10932229/82162836-bc786f00-98ea-11ea-9871-9170726f5e4c.png)

The Rpi ARM cpu is quite old, my phone (Xiaomi A2), running Andronix is faster (default blas):

Mobile/cell phone screen shot:
![image](https://user-images.githubusercontent.com/10932229/82165541-c43f1000-98f8-11ea-99f9-d4d132e8e8e8.png)

X11 screen via VNC:
![image](https://user-images.githubusercontent.com/10932229/82165622-24ce4d00-98f9-11ea-834e-7527f2c9ffd5.png)

MikhailRyazanov commented:
> Interesting that an old AMD can keep up with modern systems.

It's only about 2 years “old” (released in February 2018)... More surprising is the price — it was 99 USD (+125 for the motherboard) and now is even cheaper. At that time it had the best performance/price ratio, but I did not expect it to have this absolute single-core performance, especially in FP calculations (where AMD was always lagging).

> my phone (Xiaomi A2), running Andronix is faster (default blas)

LOL! :–) “PyAbel: real-time Abel transforms in your pocket!”
Have you tried to use its camera for data acquisition? ;–)

DanHickstein commented on doc/transform_methods/comparison/fig_benchmarks/pyabel-plot.py:
```diff
@@ -0,0 +1,107 @@
+# -*- coding: UTF-8 -*-
```
Is this file actually used any longer?

MikhailRyazanov replied:
It is not used in the final documentation but is useful for looking quickly at the results while running benchmarks (however, then it needs `benchmark_dir` to point at `'working'`, which I forgot to change). I can either delete it or modify to complement `pyabel-timing.py`. What do you think?

MikhailRyazanov replied:
Well, I have committed the usable version, just in case, but it can be deleted completely.

DanHickstein replied:
I think that it's fine to keep it around so that it can be used for looking at the benchmarks. Thanks for fixing it!

DanHickstein commented:
> @DanHickstein and @stggh, should we merge this now? 

Yeah, I think that this is looking good and can be merged. We should just make sure that we haven't included any extra files. I think one of the plotting files above is now redundant.

DanHickstein commented:
Okay, now I'm fine with merging this.

MikhailRyazanov commented:
Great! Many thanks to @DanHickstein for adapting the text and @stggh for providing the timings!
